src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Create/CreateResourceHandler.cs

```diff
@@ -54,8 +54,20 @@ public async Task<UpsertResourceResponse> Handle(CreateResourceRequest request,
 
             var resource = request.Resource.ToPoco<Resource>();
 
-            // If an Id is supplied on create it should be removed/ignored
-            resource.Id = null;
+            // In a POST request, a Transactional Bundle processing defines the ID of resources being created when a full url is provided (as part of the bundle entry).
+            // After defined, the IDs are used to update references between records in the same bundle.
+            // If the record (in the current request) is part of a transaction and had an ID assigned to it, then the ID should be preserved.
+            if (IsBundleParallelTransaction(request) && !string.IsNullOrWhiteSpace(request.BundleResourceContext?.PersistedId))
+            {
+                if (!string.Equals(resource.Id, request.BundleResourceContext.PersistedId, StringComparison.OrdinalIgnoreCase))
+                {
+                    throw new InvalidOperationException($"Resource ID mismatch: Expected '{request.BundleResourceContext.PersistedId}', but got '{resource.Id}'.");
```

Comment:
Is this strict validation a breaking change for customer?

Reply:
Can we add 'in bundle' in the error message e.g. Resource ID mismatch in bundle: ....

Reply:
Added more comments and details with the reason for this validation.